### PR TITLE
Added SSIL post processing effect

### PR DESCRIPTION
**Pull request by clayjohn**
Implements Screen Space Indirect Lighting (I still need to rename it in the code)

SSIL adds short-range indirect lighting into the scene by sampling much like SSAO. I went with the ASSAO-like approach after months of testing and experimenting with other methods (including full blown SSGI). This method strikes the best balance between performance and quality and does not require Temporal Super Sampling at all.

**SSIL is intended to complement VoxelGI or SDFGI and is not a form of GI on its own. It only adds light into the scene and it only takes lighting into account in a short to medium range**

As it stands, SSIL is slightly slower than SSAO on my hardware. Results will vary, but for me, it takes between 0.3-1.0 ms.

https://user-images.githubusercontent.com/16521339/128133354-743f3794-cd7f-4220-a4ff-1b6d1a72222b.mp4

**TODO:**

Before marking this PR as ready to merge I need to:

- [x] 1. convert all references to SSGI to SSIL
- [x] 2. clean up Screen Space reflections (likely I will just remove the changes to SSR and make them in another PR)
- [x] 3. clean up code style (there are lots of places where the code is ugly)
- [x] 4. move SSIL uniform sets into the renderbuffers
- [x] 5. ~move SSIL after AO in the main scene shader~ (nevermind, this doesn't actually make sense)
- [x] 6. write docs :)

After this PR is merged I need to investigate:
1. merging the SSIL buffer with the main GI buffer (might be nice, but limits flexibility)
2. generating Screen Space Bent Normals during SSIL pass (and using them for specular occlusion)
3. generating SSIL at quarter resolution  for performance
4. moving last frame texture copy to raster (so we avoid jumping back and forth between raster and compute

### CODE REVIEW

**Comment by clayjohn** [comment]
> Lights' Indirect Energy property is not taken into account by SSIL. If this is a limitation that can't be resolved, it should be documented in the class reference.

That's right. I've added this limitation to the docs. It will also need to be reflected in a tutorial style doc

> SSIL colors are broken (everything is grayscale) when Half Size is disabled in the Project Settings. Not that it's very practical for performance reasons, but still 🙂

I think I have fixed this. At any rate I can't reproduce on the most recent version

> Enabling MSAA while SSIL is enabled may break rendering until you look away from any SSIL'd surface and look back.

I haven't been able to reproduce this. I will keep my eye on it though. 

> It seems the scene got excessively bright when approaching the material in question. This only happens if roughness is 0 or 0.01. With a roughness of 0.02 or higher, SSIL gets quite flickery but it doesn't turn the whole scene black. Maybe this can be fixed by clamping the material roughness to a minimum value (0.15 looks good from my quick testing). 

There is certainly a risk of infinite feedback loops forming between surfaces that face each other. I should put a cap on energy to avoid this.

File: doc/classes/Environment.xml [hunk 1]
```diff
@@ -237,7 +237,22 @@
 			The distance at which objects can occlude each other when calculating screen-space ambient occlusion. Higher values will result in occlusion over a greater distance at the cost of performance and quality.
 		</member>
 		<member name="ssao_sharpness" type="float" setter="set_ssao_sharpness" getter="get_ssao_sharpness" default="0.98">
-			Sharpness refers to the amount that the screen-space ambient occlusion effect is allowed to blur over the edges of objects. Setting too high will result in aliasing around the edges of objects. Setting too low will make object edges appear blurry.
+			The amount that the screen-space ambient occlusion effect is allowed to blur over the edges of objects. Setting too high will result in aliasing around the edges of objects. Setting too low will make object edges appear blurry.
```
Comment (mrjustaguy) [inline doc/classes/Environment.xml:240]: ambient occlusion -> indirect lighting

File: doc/classes/ProjectSettings.xml [hunk 2]
```diff
@@ -1551,6 +1551,24 @@
...
 		</member>
+		<member name="rendering/environment/ssil/adaptive_target" type="float" setter="" getter="" default="0.5">
+			Quality target to use when [member rendering/environment/ssil/quality] is set to [code]ULTRA[/code]. A value of [code]0.0[/code] provides a quality and speed similar to [code]MEDIUM[/code] while a value of [code]1.0[/code] provides much higher quality than any of the other settings at the cost of performance.
+		</member>
+		<member name="rendering/environment/ssil/blur_passes" type="int" setter="" getter="" default="4">
+			Number of blur passes to use when computing screen-space indirect lighting. A higher number will result in a smoother look, but will be slower to compute and will have less high-frequency detail.
+		</member>
+		<member name="rendering/environment/ssil/fadeout_from" type="float" setter="" getter="" default="50.0">
+			Distance at which the screen-space ambient occlusion effect starts to fade out. Use this hide screen-space indirect lighting at great distances.
```
Comment (mrjustaguy) [inline doc/classes/ProjectSettings.xml:1561]: ambient occlusion -> indirect lighting

File: servers/rendering/renderer_rd/effects_rd.h [hunk 3]
```diff
@@ -489,10 +508,104 @@ class EffectsRD {
...
+		float intensity;
+		int size_multiplier;
+		int pad;
+
+		float fade_out_mul;
+		float fade_out_add;
+		float normal_rejection_amount;
+		float inv_radius_near_limit;
+
+		bool is_orthogonal;
```
Comment (reduz) [inline servers/rendering/renderer_rd/effects_rd.h:555]: Remember not use bool in push constants, some compilers will mess up alignment. Always use uint32_t.

**Review by reduz — CHANGES_REQUESTED** [review]

**Comment by clayjohn** [comment]
Okay updated with fixes for all the identified issues. I have solved the issue with propagating INFs and NaNs. It should be ready to go.

**Comment by clayjohn** [comment]
After spending more time thinking about this, I think I can polish it a bit more before it is ready to merge.

**Comment by Calinou** [comment]
It seems enabling SSIL creates an SSAO-like effect automatically. Should SSAO be disabled when SSIL is enabled?

**Comment by mrjustaguy** [comment]
Eh, I know that technically SSIL on Ultra is broken (as it actually lights up a lightless scene on higher intensity, and also behaves differently to all the other settings), but I really love the Ultra setting effect, just have to set 0.1 intensity instead of 1 for the scene not to be lit by it without lights in a noticable way...

**Comment by clayjohn** [comment]
Just pushed documentation updates to address recent comments. If CI is happy, I think this is ready for final review and merging!

**Comment by mrjustaguy** [comment]
Hmm, Could SSIL maybe use Indirect Lighting by adding an extra step, instead of just using Direct Lighting to contribute to the end result?

I mean the way I see it, there are a few different possible implementations that would allow for Indirect Energy from lights to contribute to SSIL instead of just Energy. Here are some ideas

1) Compute a Frame where you Determine which pixel is affected by which lights, and then use the Indirect energy to multiply the output of the end result Light contribution by that amount
2) Compute an Extra Frame where instead of using Energy for the Light's energy, you use Indirect Energy, and mix the two results
3) Just render a Frame with Indirect Lighting Energy Instead of Direct Lighting Energy, and pass that to SSIL instead of the current Direct Lighting Energy that's passed to it.

I mean this could help with giving more fine-tune control over the intensity of the effect on a Per-Light basis, which would be great for artistic control

**Comment by unfa** [comment]
In some cases (usually with exagerrated settings) the geometry wireframe becomes visible. I wonder what could be causing this:
image

It's exagerrated when sharpness is turned up:
image

Of course this example  uses an extremely rage radius.

It seems to occur only from Medium quality upwards.

On Low:
image

On Medium:
image

**Comment by Calinou** [comment]
> It seems to occur only from Medium quality upwards.

The wireframe is present on Low too, but it's hard to discern:

image

**Comment by unfa** [comment]
BTW, I'm testing what's the limit:
https://youtu.be/O39GyzSdZfc

It seems that if emissive sources are touching the screen border - sudden light energy increase is produced.
Maybe that could be tweaked?

**Comment by mrjustaguy** [comment]
Is the Scene Geometry CSG Nodes? They Try disabling all gizmos and see if the geometry is still visible, or run the project..
I've noticed that it incorrectly uses in-Editor stuff in it's results, such as Skeletons, CSG node wireframes, etc...

**Comment by unfa** [comment]
No, the scene geometry is imported from a GLTF exported from Blender. No CSG is used.

**Comment by clayjohn** [comment]
@unfa The "wireframe" artifacts look like there may be an issue with your mesh normals along triangle edges. What does it look like when you use the "normal buffer" render mode from the perspective menu?

**Comment by unfa** [comment]
The environment normal pass:
image

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Rebased locally to make sure that it still builds fine.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**